### PR TITLE
Add regexp.replace support in role templates

### LINKED ISSUE #3374: Extend variable interpolation syntax
## Description

Current role variables only support the following sytnax:

```yaml
kuberentes_groups: ['{{variable_name}}']
```

with the proposal implemented, the following will become possible:

```yaml
kuberentes_groups: ['IAM#{{variable_name}}-suffix']
```

In addition function calls will be possible with addition of `regexp_replace` function:

```yaml
logins: ['{{regexp.replace(user.spec.traits[email], "^(.)@.*, "\1")}}']
```

Which is consistent with our current variable `where` syntax:

https://gravitational.com/gravity/docs/hub/#cluster-rbac-using-labels

```yaml
where: contains(user.spec.traits["roles"], resource.metadata.labels["team"])
```

## Functions

Two functions will be added:

Generic regexp replace function for common cases:

```yaml
regexp.replace(what, expression, replacement)
```

Specific function to handle a very common edge case for capturing local part of the email:

```yaml
email.local(user.spec.traits["email"])
```

## Properties access

Our simplified version supported `external` and `internal` shortcuts. The extended version supports
access to arbitrary metadata of the user context:

```yaml
# access to user traits dictionary
user.spec.traits
# access to user email trait
user.spec.traits["email"]
# access to cluster metadata labels
resource.metadata.labels
```
```

### PR DESCRIPTION
**Pull request by nklaassen**
Closes #3374 

This PR adds support for `{{regexp.replace(namespace.variable, expression, replacement)}}` syntax in role templates.

See the updated tests for some examples.

### CODE REVIEW

**Comment by russjones** [comment]
@stevenGravy Can you verify this solves the customers issue, I think they are using `regexp.replace` in Kubernetes Groups in their role?

**Review by klizhentas — APPROVED** [review]
@nklaassen looks good UX-wise, but need docs in this PR please

**Comment by nklaassen** [comment]
> @nklaassen looks good UX-wise, but need docs in this PR please

@klizhentas sure, https://goteleport.com/docs/access-controls/guides/role-templates/#interpolation-rules seems like a good place to add an example, I'll do that.

@klizhentas @russjones what should the behaviour be when the expression does not match at all? As-is it will just pass the value through unchanged, but I'm realizing that may not be what we want.

I'm thinking of a case like `{{regexp.replace(external.email, "(.*)@example1.com", "$1")}}`, but the input (`external.email`) is `alice@example2.com`. In this case should it be an error and return nothing, or pass `alice@example2.com` through unchanged?

**Comment by nklaassen** [comment]
In the latest commit, all values which do not match the regular expression at all are filtered out

File: docs/pages/access-controls/guides/role-templates.mdx [hunk 1]
```diff
@@ -281,6 +281,8 @@ spec:
 
     # Functions transform variables.
     database_users: ['{{email.local(external.email)}}']
+    database_labels:
+      'env': '{{regexp.replace(external.acces["env"], "^(staging)$", "$1")}}']
```
Comment (inertial-frame) [inline docs/pages/access-controls/guides/role-templates.mdx:285]: Is there a dangling `]` here?

Comment (nklaassen) [reply docs/pages/access-controls/guides/role-templates.mdx]: Yeah my bad, a couple typos slipped through here actually, should be fixed in the latest commit

**Comment by russjones** [comment]
@nklaassen Once this is merged, please backport to `branch/v6.2`.

**Review by tcsc — APPROVED** [review]

File: lib/utils/parse/parse.go [hunk 2]
```diff
@@ -69,6 +69,34 @@ func (emailLocalTransformer) transform(in string) (string, error) {
...
+	}
+	return &regexpReplaceTransformer{
+		re:          re,
+		replacement: replacement,
+	}, nil
+}
+
+// transform applies the regexp replacement (with expansion)
+func (r regexpReplaceTransformer) transform(in string) (string, error) {
+	// filter out inputs which to not match the regexp at all
```
Comment (r0mant) [inline lib/utils/parse/parse.go:93]: to not -> do not

File: lib/utils/parse/parse.go [hunk 3]
```diff
@@ -69,6 +69,34 @@ func (emailLocalTransformer) transform(in string) (string, error) {
...
+	if !r.re.MatchString(in) {
+		return "", nil
```
Comment (r0mant) [inline lib/utils/parse/parse.go:95]: Shouldn't this return `in` unmodified?

**Review by r0mant — APPROVED** [review]

**Comment by klizhentas** [comment]
@nklaassen  filtering out non matching values is a good default for both allow and deny expressions

File: lib/utils/parse/parse.go (hunk 3 above)
Comment (nklaassen) [reply lib/utils/parse/parse.go:95]: I'm filtering out inputs which do not match the regex, that seems to be more useful and the behaviour people actually want. Empty strings are removed from the output set in `Interpolate`.

**Pull request merged**